File: apps/api/src/modules/printing/types/print-document.ts
export const PRINT_DOCUMENT_SCHEMA_V2 = 2 as const;
export const PRINT_DOCUMENT_SCHEMA_V3 = 3 as const;
export const PRINT_DOCUMENT_SCHEMA_VERSION = PRINT_DOCUMENT_SCHEMA_V2;
export const PRINT_DOCUMENT_TYPES = ['PRINT_DOCUMENT'] as const;
export const PRINT_PAPER_WIDTHS = ['MM58', 'MM80'] as const;
export const PRINT_ALIGNMENTS = ['LEFT', 'CENTER', 'RIGHT'] as const;
export const PRINT_FONT_SIZES = ['SMALL', 'NORMAL', 'LARGE'] as const;
export const PRINT_CUT_MODES = ['NONE', 'HALF', 'FULL'] as const;
export const PRINT_COLUMN_OVERFLOWS = ['ELLIPSIS', 'FIT'] as const;

export type TextPrintBlock = {
  type: 'TEXT';
  text: string;
  align: (typeof PRINT_ALIGNMENTS)[number];
  bold: boolean;
  fontSize: (typeof PRINT_FONT_SIZES)[number];
  underline: boolean;
};

export type TextPrintBlockV3 = TextPrintBlock & {
  overflow?: (typeof PRINT_COLUMN_OVERFLOWS)[number];
};

export type RowPrintBlock = { type: 'ROW'; left: string; right: string; bold: boolean };
export type DividerPrintBlock = { type: 'DIVIDER' };
export type FeedPrintBlock = { type: 'FEED'; lines: number };
export type CutPrintBlock = { type: 'CUT'; mode: (typeof PRINT_CUT_MODES)[number] };

export type PrintBlock =
  | TextPrintBlock
  | RowPrintBlock
  | DividerPrintBlock
  | FeedPrintBlock
  | CutPrintBlock;

export interface PrintColumnCell {
  text: string;
  weight: number;
  align: (typeof PRINT_ALIGNMENTS)[number];
  bold: boolean;
  fontSize: (typeof PRINT_FONT_SIZES)[number];
  overflow: (typeof PRINT_COLUMN_OVERFLOWS)[number];
  paddingDots: number;
}

export type ColumnsPrintBlock = {
  type: 'COLUMNS';
  gapDots: number;
  cells: PrintColumnCell[];
};

export type BoxedTitlePrintBlock = {
  type: 'BOXED_TITLE';
  boxText: string;
  title: string;
  subtitle: string;
  boxWeight: number;
  gapDots: number;
  fontSize: (typeof PRINT_FONT_SIZES)[number];
};

export type PrintBlockV3 =
  | Exclude<PrintBlock, TextPrintBlock>
  | TextPrintBlockV3
  | ColumnsPrintBlock
  | BoxedTitlePrintBlock;

/**
 * Presentation-only print protocol. It intentionally contains no order,
 * settlement, discount, rounding, item-category, or receipt-type semantics.
 */
export interface PrintDocumentV2 {
  documentType: 'PRINT_DOCUMENT';
  schemaVersion: 2;
  paperWidth: (typeof PRINT_PAPER_WIDTHS)[number];
  copies: number;
  blocks: PrintBlock[];
}

/** Schema 3 retains every V2 block and adds measured pixel-column primitives. */
export interface PrintDocumentV3 {
  documentType: 'PRINT_DOCUMENT';
  schemaVersion: 3;
  paperWidth: (typeof PRINT_PAPER_WIDTHS)[number];
  copies: number;
  blocks: PrintBlockV3[];
}

export type PrintDocument = PrintDocumentV2 | PrintDocumentV3;

export function isPrintDocumentV2(value: unknown): value is PrintDocumentV2 {
  return isPrintDocumentVersion(value, PRINT_DOCUMENT_SCHEMA_V2);
}

export function isPrintDocumentV3(value: unknown): value is PrintDocumentV3 {
  return isPrintDocumentVersion(value, PRINT_DOCUMENT_SCHEMA_V3);
}

export function isPrintDocument(value: unknown): value is PrintDocument {
  return isPrintDocumentV2(value) || isPrintDocumentV3(value);
}

export function assertPrintDocumentV2(value: unknown): asserts value is PrintDocumentV2 {
  assertPrintDocument(value, PRINT_DOCUMENT_SCHEMA_V2);
}

export function assertPrintDocumentV3(value: unknown): asserts value is PrintDocumentV3 {
  assertPrintDocument(value, PRINT_DOCUMENT_SCHEMA_V3);
}

function isPrintDocumentVersion(value: unknown, schemaVersion: 2 | 3) {
  return (
    isPlainObject(value) &&
    value.documentType === 'PRINT_DOCUMENT' &&
    value.schemaVersion === schemaVersion
  );
}

function assertPrintDocument(
  value: unknown,
  schemaVersion: 2 | 3,
): asserts value is PrintDocument {
  if (!isPlainObject(value) || !hasOnlyKeys(value, [
    'documentType', 'schemaVersion', 'paperWidth', 'copies', 'blocks',
  ])) {
    throw new Error('Print document must be a supported object');
  }
  if (
    value.documentType !== 'PRINT_DOCUMENT' ||
    value.schemaVersion !== schemaVersion ||
    !PRINT_PAPER_WIDTHS.includes(value.paperWidth as never) ||
    !Number.isSafeInteger(value.copies) ||
    Number(value.copies) < 1 ||
    Number(value.copies) > 10 ||
    !Array.isArray(value.blocks) ||
    value.blocks.length < 1 ||
    value.blocks.length > 2_000
  ) {
    throw new Error(`Print document does not match schema version ${schemaVersion}`);
  }
  value.blocks.forEach((block, index) => assertBlock(block, index, schemaVersion));
  const cutIndexes = value.blocks
    .map((block, index) => isPlainObject(block) && block.type === 'CUT' ? index : -1)
    .filter((index) => index >= 0);
  if (cutIndexes.length > 1 || (cutIndexes.length === 1 && cutIndexes[0] !== value.blocks.length - 1)) {
    throw new Error('CUT must be the final and only CUT block');
  }
}

function assertBlock(value: unknown, index: number, schemaVersion: 2 | 3) {
  if (!isPlainObject(value) || typeof value.type !== 'string') {
    throw new Error(`Print block ${index} is invalid`);
  }
  switch (value.type) {
    case 'TEXT':
      if (
        !hasOnlyKeys(value, schemaVersion === PRINT_DOCUMENT_SCHEMA_V3
          ? ['type', 'text', 'align', 'bold', 'fontSize', 'underline', 'overflow']
          : ['type', 'text', 'align', 'bold', 'fontSize', 'underline']) ||
        !boundedText(value.text, 0, 2_000) ||
        !PRINT_ALIGNMENTS.includes(value.align as never) ||
        typeof value.bold !== 'boolean' ||
        !PRINT_FONT_SIZES.includes(value.fontSize as never) ||
        typeof value.underline !== 'boolean' ||
        (value.overflow !== undefined && !PRINT_COLUMN_OVERFLOWS.includes(value.overflow as never))
      ) throw new Error(`TEXT block ${index} is invalid`);
      return;
    case 'ROW':
      if (
        !hasOnlyKeys(value, ['type', 'left', 'right', 'bold']) ||
        !boundedText(value.left, 0, 1_000) ||
        !boundedText(value.right, 0, 1_000) ||
        typeof value.bold !== 'boolean'
      ) throw new Error(`ROW block ${index} is invalid`);
      return;
    case 'DIVIDER':
      if (!hasOnlyKeys(value, ['type'])) throw new Error(`DIVIDER block ${index} is invalid`);
      return;
    case 'FEED':
      if (
        !hasOnlyKeys(value, ['type', 'lines']) ||
        !Number.isSafeInteger(value.lines) ||
        !inRange(Number(value.lines), 1, 20)
      ) {
        throw new Error(`FEED block ${index} is invalid`);
      }
      return;
    case 'CUT':
      if (
        !hasOnlyKeys(value, ['type', 'mode']) ||
        !PRINT_CUT_MODES.includes(value.mode as never)
      ) {
        throw new Error(`CUT block ${index} is invalid`);
      }
      return;
    case 'COLUMNS':
      if (schemaVersion !== PRINT_DOCUMENT_SCHEMA_V3) {
        throw new Error(`Print block ${index} type is unsupported`);
      }
      assertColumnsBlock(value, index);
      return;
    case 'BOXED_TITLE':
      if (schemaVersion !== PRINT_DOCUMENT_SCHEMA_V3) {
        throw new Error(`Print block ${index} type is unsupported`);
      }
      if (
        !hasOnlyKeys(value, [
          'type', 'boxText', 'title', 'subtitle', 'boxWeight', 'gapDots', 'fontSize',
        ]) ||
        !boundedText(value.boxText, 1, 64) ||
        !boundedText(value.title, 1, 200) ||
        !boundedText(value.subtitle, 1, 64) ||
        !Number.isSafeInteger(value.boxWeight) ||
        !inRange(Number(value.boxWeight), 10, 50) ||
        !Number.isSafeInteger(value.gapDots) ||
        !inRange(Number(value.gapDots), 0, 40) ||
        !PRINT_FONT_SIZES.includes(value.fontSize as never)
      ) {
        throw new Error(`BOXED_TITLE block ${index} is invalid`);
      }
      return;
    default:
      throw new Error(`Print block ${index} type is unsupported`);
  }
}

function assertColumnsBlock(value: Record<string, unknown>, index: number) {
  if (
    !hasOnlyKeys(value, ['type', 'gapDots', 'cells']) ||
    !Number.isSafeInteger(value.gapDots) ||
    !inRange(Number(value.gapDots), 0, 40) ||
    !Array.isArray(value.cells) ||
    !inRange(value.cells.length, 2, 4)
  ) {
    throw new Error(`COLUMNS block ${index} is invalid`);
  }
  value.cells.forEach((cell, cellIndex) => {
    if (
      !isPlainObject(cell) ||
      !hasOnlyKeys(cell, [
        'text', 'weight', 'align', 'bold', 'fontSize', 'overflow', 'paddingDots',
      ]) ||
      !boundedText(cell.text, 0, 2_000) ||
      !Number.isSafeInteger(cell.weight) ||
      !inRange(Number(cell.weight), 1, 100) ||
      !PRINT_ALIGNMENTS.includes(cell.align as never) ||
      typeof cell.bold !== 'boolean' ||
      !PRINT_FONT_SIZES.includes(cell.fontSize as never) ||
      !PRINT_COLUMN_OVERFLOWS.includes(cell.overflow as never) ||
      !Number.isSafeInteger(cell.paddingDots) ||
      !inRange(Number(cell.paddingDots), 0, 24)
    ) {
      throw new Error(`COLUMNS block ${index} cell ${cellIndex} is invalid`);
    }
  });
}

function boundedText(value: unknown, min: number, max: number) {
  return typeof value === 'string' && value.length >= min && value.length <= max && !value.includes('\0');
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasOnlyKeys(value: Record<string, unknown>, allowed: string[]) {
  const allowedSet = new Set(allowed);
  return Object.keys(value).every((key) => allowedSet.has(key));
}

function inRange(value: number, min: number, max: number) {
  return value >= min && value <= max;
}

File: clients/windows/YunQiao.Cashier/src/YunQiao.Cashier.Core/Protocol/PrintDocument.cs
using System.Text.Json;

namespace YunQiao.Cashier.Core.Protocol;

public enum PrintPaperWidth { MM58, MM80 }
public enum PrintAlignment { LEFT, CENTER, RIGHT }
public enum PrintFontSize { SMALL, NORMAL, LARGE }
public enum PrintCutMode { NONE, HALF, FULL }
public enum PrintColumnOverflow { ELLIPSIS, FIT }

public sealed record PrintColumnCell(
    string Text,
    int Weight,
    PrintAlignment Align,
    bool Bold,
    PrintFontSize FontSize,
    PrintColumnOverflow Overflow,
    int PaddingDots);

public abstract record PrintBlock
{
    public sealed record Text(
        string Value,
        PrintAlignment Align,
        bool Bold,
        PrintFontSize FontSize,
        bool Underline,
        PrintColumnOverflow? Overflow) : PrintBlock;

    public sealed record Row(string Left, string Right, bool Bold) : PrintBlock;
    public sealed record Columns(int GapDots, IReadOnlyList<PrintColumnCell> Cells) : PrintBlock;
    public sealed record BoxedTitle(
        string BoxText,
        string Title,
        string Subtitle,
        int BoxWeight,
        int GapDots,
        PrintFontSize FontSize) : PrintBlock;
    public sealed record Divider : PrintBlock;
    public sealed record Feed(int Lines) : PrintBlock;
    public sealed record Cut(PrintCutMode Mode) : PrintBlock;
}

public sealed record PrintDocument(
    int SchemaVersion,
    PrintPaperWidth PaperWidth,
    int Copies,
    IReadOnlyList<PrintBlock> Blocks)
{
    public int WidthDots => PaperWidth == PrintPaperWidth.MM58 ? 384 : 576;

    public PrintCutMode CutMode => Blocks.LastOrDefault() is PrintBlock.Cut cut
        ? cut.Mode
        : PrintCutMode.NONE;
}

/// <summary>Strict port of Android PrintDocumentV2Parser. Business fields and raw commands are rejected.</summary>
public static class PrintDocumentParser
{
    public const int MaxJsonChars = 512_000;

    public static PrintDocument Parse(string json)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(json);
        if (json.Length is < 2 or > MaxJsonChars)
            throw new PrintDocumentException("Print document size is invalid.");

        using var parsed = JsonDocument.Parse(json, new JsonDocumentOptions
        {
            AllowTrailingCommas = false,
            CommentHandling = JsonCommentHandling.Disallow,
            MaxDepth = 64,
        });
        var root = parsed.RootElement;
        RequireObject(root, "$");
        RequireOnly(root, "$", "documentType", "schemaVersion", "paperWidth", "copies", "blocks");
        if (RequiredText(root, "documentType", 32) != "PRINT_DOCUMENT")
            throw new PrintDocumentException("Unsupported print document type.");

        var schema = RequiredInt(root, "schemaVersion", 2, 3);
        var paper = RequiredEnum<PrintPaperWidth>(root, "paperWidth", 8);
        var copies = RequiredInt(root, "copies", 1, 10);
        if (!root.TryGetProperty("blocks", out var blocksJson) || blocksJson.ValueKind != JsonValueKind.Array)
            throw new PrintDocumentException("Print blocks are missing.");
        if (blocksJson.GetArrayLength() is < 1 or > 2_000)
            throw new PrintDocumentException("Print block count is invalid.");

        var blocks = new List<PrintBlock>(blocksJson.GetArrayLength());
        var index = 0;
        foreach (var item in blocksJson.EnumerateArray())
        {
            var path = $"$.blocks[{index}]";
            RequireObject(item, path);
            var type = RequiredText(item, "type", 16);
            blocks.Add(type switch
            {
                "TEXT" => ParseText(item, path, schema),
                "ROW" => ParseRow(item, path),
                "COLUMNS" when schema == 3 => ParseColumns(item, path),
                "BOXED_TITLE" when schema == 3 => ParseBoxedTitle(item, path),
                "DIVIDER" => ParseDivider(item, path),
                "FEED" => ParseFeed(item, path),
                "CUT" => ParseCut(item, path),
                "COLUMNS" or "BOXED_TITLE" => throw new PrintDocumentException($"{type} requires print document schema 3."),
                _ => throw new PrintDocumentException($"Unsupported print block type: {type}"),
            });
            index++;
        }

        var cutIndexes = blocks.Select((value, blockIndex) => (value, blockIndex))
            .Where(value => value.value is PrintBlock.Cut)
            .Select(value => value.blockIndex)
            .ToArray();
        if (cutIndexes.Length > 1) throw new PrintDocumentException("Only one CUT block is supported.");
        if (cutIndexes.Length == 1 && cutIndexes[0] != blocks.Count - 1)
            throw new PrintDocumentException("CUT must be the final print block.");

        return new PrintDocument(schema, paper, copies, blocks);
    }

    private static PrintBlock ParseText(JsonElement value, string path, int schema)
    {
        RequireOnly(value, path, schema == 3
            ? ["type", "text", "align", "bold", "fontSize", "underline", "overflow"]
            : ["type", "text", "align", "bold", "fontSize", "underline"]);
        PrintColumnOverflow? overflow = null;
        if (schema == 3 && value.TryGetProperty("overflow", out var overflowJson))
            overflow = ParseEnum<PrintColumnOverflow>(RequiredStringValue(overflowJson, "overflow", 16));
        return new PrintBlock.Text(
            RequiredTextAllowEmpty(value, "text", 2_000),
            RequiredEnum<PrintAlignment>(value, "align", 16),
            RequiredBoolean(value, "bold"),
            RequiredEnum<PrintFontSize>(value, "fontSize", 16),
            RequiredBoolean(value, "underline"),
            overflow);
    }

    private static PrintBlock ParseRow(JsonElement value, string path)
    {
        RequireOnly(value, path, "type", "left", "right", "bold");
        return new PrintBlock.Row(
            RequiredTextAllowEmpty(value, "left", 1_000),
            RequiredTextAllowEmpty(value, "right", 1_000),
            RequiredBoolean(value, "bold"));
    }

    private static PrintBlock ParseColumns(JsonElement value, string path)
    {
        RequireOnly(value, path, "type", "gapDots", "cells");
        if (!value.TryGetProperty("cells", out var cellsJson) || cellsJson.ValueKind != JsonValueKind.Array)
            throw new PrintDocumentException("Column cells are missing.");
        if (cellsJson.GetArrayLength() is < 2 or > 4)
            throw new PrintDocumentException("Column cell count is invalid.");
        var cells = new List<PrintColumnCell>(cellsJson.GetArrayLength());
        var index = 0;
        foreach (var cell in cellsJson.EnumerateArray())
        {
            var cellPath = $"{path}.cells[{index}]";
            RequireObject(cell, cellPath);
            RequireOnly(cell, cellPath, "text", "weight", "align", "bold", "fontSize", "overflow", "paddingDots");
            cells.Add(new PrintColumnCell(
                RequiredTextAllowEmpty(cell, "text", 2_000),
                RequiredInt(cell, "weight", 1, 100),
                RequiredEnum<PrintAlignment>(cell, "align", 16),
                RequiredBoolean(cell, "bold"),
                RequiredEnum<PrintFontSize>(cell, "fontSize", 16),
                RequiredEnum<PrintColumnOverflow>(cell, "overflow", 16),
                RequiredInt(cell, "paddingDots", 0, 24)));
            index++;
        }
        return new PrintBlock.Columns(RequiredInt(value, "gapDots", 0, 40), cells);
    }

    private static PrintBlock ParseBoxedTitle(JsonElement value, string path)
    {
        RequireOnly(value, path, "type", "boxText", "title", "subtitle", "boxWeight", "gapDots", "fontSize");
        return new PrintBlock.BoxedTitle(
            RequiredText(value, "boxText", 64),
            RequiredText(value, "title", 200),
            RequiredText(value, "subtitle", 64),
            RequiredInt(value, "boxWeight", 10, 50),
            RequiredInt(value, "gapDots", 0, 40),
            RequiredEnum<PrintFontSize>(value, "fontSize", 16));
    }

    private static PrintBlock ParseDivider(JsonElement value, string path)
    {
        RequireOnly(value, path, "type");
        return new PrintBlock.Divider();
    }

    private static PrintBlock ParseFeed(JsonElement value, string path)
    {
        RequireOnly(value, path, "type", "lines");
        return new PrintBlock.Feed(RequiredInt(value, "lines", 1, 20));
    }

    private static PrintBlock ParseCut(JsonElement value, string path)
    {
        RequireOnly(value, path, "type", "mode");
        return new PrintBlock.Cut(RequiredEnum<PrintCutMode>(value, "mode", 8));
    }

    private static void RequireObject(JsonElement value, string path)
    {
        if (value.ValueKind != JsonValueKind.Object)
            throw new PrintDocumentException($"Print document value at {path} must be an object.");
    }

    private static void RequireOnly(JsonElement value, string path, params string[] allowed)
    {
        var accepted = allowed.ToHashSet(StringComparer.Ordinal);
        var unsupported = value.EnumerateObject().Select(property => property.Name)
            .Where(name => !accepted.Contains(name)).Order(StringComparer.Ordinal).ToArray();
        if (unsupported.Length > 0)
            throw new PrintDocumentException($"Print document contains unsupported fields at {path}: {string.Join(',', unsupported)}.");
    }

    private static string RequiredText(JsonElement owner, string key, int max) =>
        owner.TryGetProperty(key, out var value) ? RequiredStringValue(value, key, max) : throw Invalid(key);

    private static string RequiredStringValue(JsonElement value, string key, int max)
    {
        var text = RequiredStringValueAllowEmpty(value, key, max);
        if (text.Length == 0) throw Invalid(key);
        return text;
    }

    private static string RequiredTextAllowEmpty(JsonElement owner, string key, int max) =>
        owner.TryGetProperty(key, out var value) ? RequiredStringValueAllowEmpty(value, key, max) : throw Invalid(key);

    private static string RequiredStringValueAllowEmpty(JsonElement value, string key, int max)
    {
        if (value.ValueKind != JsonValueKind.String) throw Invalid(key);
        var text = value.GetString() ?? string.Empty;
        if (text.Length > max || text.Contains('\0')) throw Invalid(key);
        return text;
    }

    private static bool RequiredBoolean(JsonElement owner, string key)
    {
        if (!owner.TryGetProperty(key, out var value) || value.ValueKind is not (JsonValueKind.True or JsonValueKind.False))
            throw Invalid(key);
        return value.GetBoolean();
    }

    private static int RequiredInt(JsonElement owner, string key, int minimum, int maximum)
    {
        if (!owner.TryGetProperty(key, out var value) || value.ValueKind != JsonValueKind.Number || !value.TryGetInt32(out var number))
            throw Invalid(key);
        if (number < minimum || number > maximum) throw new PrintDocumentException($"Print integer is out of range: {key}");
        return number;
    }

    private static T RequiredEnum<T>(JsonElement owner, string key, int max) where T : struct, Enum =>
        ParseEnum<T>(RequiredText(owner, key, max));

    private static T ParseEnum<T>(string value) where T : struct, Enum =>
        Enum.TryParse<T>(value, ignoreCase: false, out var parsed) && Enum.IsDefined(parsed)
            ? parsed
            : throw new PrintDocumentException($"Invalid {typeof(T).Name}: {value}");

    private static PrintDocumentException Invalid(string key) => new($"Invalid print field: {key}");
}

public sealed class PrintDocumentException(string message) : Exception(message);
